File: src/flash-api.ts
import { raiseForCode } from "./errors.js";

// TODO(release): 上线前换回 https://mineru.net/api/v1/agent
const DEFAULT_FLASH_BASE_URL = "https://staging.mineru.org.cn/api/v1/agent";

interface FlashApiResponse {
  code: number;
  msg?: string;
  trace_id?: string;
  data: Record<string, unknown>;
}

export { DEFAULT_FLASH_BASE_URL };

export class FlashApiClient {
  private readonly baseUrl: string;
  private readonly headers: Record<string, string>;
  private source: string;

  constructor(baseUrl: string = DEFAULT_FLASH_BASE_URL, source = "") {
    this.baseUrl = baseUrl;
    this.headers = { "Content-Type": "application/json" };
    this.source = source;
  }

  setSource(source: string): void {
    this.source = source;
  }

  async post(
    path: string,
    json: Record<string, unknown>,
  ): Promise<FlashApiResponse> {
    const headers: Record<string, string> = { ...this.headers };
    if (this.source) {
      headers["source"] = this.source;
    }
    const resp = await fetch(`${this.baseUrl}${path}`, {
      method: "POST",
      headers,
      body: JSON.stringify(json),
    });
    return this.handle(resp);
  }

  async get(path: string): Promise<FlashApiResponse> {
    const resp = await fetch(`${this.baseUrl}${path}`, {
      method: "GET",
    });
    return this.handle(resp);
  }

  async putFile(url: string, data: Uint8Array): Promise<void> {
    const resp = await fetch(url, { method: "PUT", body: data });
    if (!resp.ok) {
      throw new Error(`Upload failed: ${resp.status} ${resp.statusText}`);
    }
  }

  async downloadText(url: string): Promise<string> {
    const resp = await fetch(url, { redirect: "follow" });
    if (!resp.ok) {
      throw new Error(`Download failed: ${resp.status} ${resp.statusText}`);
    }
    return resp.text();
  }

  private async handle(resp: Response): Promise<FlashApiResponse> {
    if (resp.status === 429) {
      raiseForCode(
        "RATE_LIMITED",
        "flash API rate limit exceeded; try again later",
      );
    }
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(
        `HTTP ${resp.status}: ${resp.statusText}${text ? ` — ${text}` : ""}`,
      );
    }
    const body = (await resp.json()) as FlashApiResponse;
    if (body.code !== 0) {
      raiseForCode(body.code, body.msg ?? "unknown error", body.trace_id ?? "");
    }
    return body;
  }
}

File: sdk/python/src/mineru/_flash_api.py
"""Low-level HTTP wrapper for the Flash (agent) API."""

from __future__ import annotations

from typing import Any

import httpx

from ._constants import DEFAULT_FLASH_BASE_URL
from .exceptions import raise_for_code

_TIMEOUT = httpx.Timeout(30.0, read=120.0)


class FlashApiClient:
    """No-auth HTTP client for the Flash API."""

    def __init__(self, base_url: str = DEFAULT_FLASH_BASE_URL, source: str = "") -> None:
        self._source = source
        self._client = httpx.Client(
            base_url=base_url,
            headers={"Content-Type": "application/json"},
            timeout=_TIMEOUT,
        )

    @property
    def source(self) -> str:
        return self._source

    @source.setter
    def source(self, value: str) -> None:
        self._source = value

    def close(self) -> None:
        self._client.close()

    def post(self, path: str, json: dict[str, Any]) -> dict[str, Any]:
        headers = {"source": self._source} if self._source else {}
        resp = self._client.post(path, json=json, headers=headers)
        return self._handle(resp)

    def get(self, path: str) -> dict[str, Any]:
        resp = self._client.get(path)
        return self._handle(resp)

    def put_file(self, url: str, data: bytes) -> None:
        resp = httpx.put(url, content=data, timeout=_TIMEOUT)
        resp.raise_for_status()

    def download_text(self, url: str) -> str:
        resp = httpx.get(url, timeout=httpx.Timeout(30.0, read=300.0), follow_redirects=True)
        resp.raise_for_status()
        return resp.text

    @staticmethod
    def _handle(resp: httpx.Response) -> dict[str, Any]:
        if resp.status_code == 429:
            raise_for_code("RATE_LIMITED", "flash API rate limit exceeded; try again later")
        resp.raise_for_status()
        body: dict[str, Any] = resp.json()
        code = body.get("code", 0)
        if code != 0:
            raise_for_code(code, body.get("msg", "unknown error"), body.get("trace_id", ""))
        return body
